lib/OpenQA/WebAPI/Plugin/ObsRsync/Gru.pm

```diff
@@ -0,0 +1,179 @@
+#
+# Normal reply should be STARTED, meaning that job shall run immediatelly or soon
+# (honouring `concurrency` configuration). (I.e. job `obs_rsync_run` has been created).
+#
+# If new request comes for a project which already has rsync running - the response will be QUEUED,
+# meaning that rsync for the project will be attempted again once active rsync is finished.
+# (I.e. job `obs_rsync_queue` is created, which later will create job `obs_rsync_run`).
+#
+# Response IN_QUEUE means that no new job was created, because the project has another job already scheduled.
+# IN_QUEUE may happen in two cases:
+# * the project has job which is already QUEUED (i.e. has task `obs_rsync_queue`)
+# * the project is STARTED, but is waiting for concurrency slot
+# (i.e. has task `obs_rsync_run` with note {waitingconcurrencyslot=>1})
+#
+# QUEUE_FULL means that number of non-finished jobs for tasks 'obs_rsync_queue' and 'obs_rsync_run'
+# reached parameter `queue_limit`.
+# This condition shouldn't practically happen and means either overload or some bug in queueing.
+#
+# Code below does not have strong concurrency guards, but in worst case it should lead to project queued
+# twice, which shouldn't be a problem.
+use constant {
+    QUEUED     => 200,
+    STARTED    => 201,
+    IN_QUEUE   => 208,
+    QUEUE_FULL => 507,
+};
+
+my $lock_timeout = 36000;
+
+sub register_tasks {
+    my ($app) = @_;
+    my $home = $app->config->{obs_rsync}->{home};
+    return undef unless $home;
+
+    my $minion = $app->minion;
+    $minion->add_task(obs_rsync_queue => sub { return _queue($app, @_) });
+    $minion->add_task(obs_rsync_run   => sub { return _run($app, @_) });
+}
+
+sub index {
+    my $self = shift;
+    my %jobs;
+    my $results = $self->app->minion->backend->list_jobs(0, 400,
+        {tasks => ['obs_rsync_run', 'obs_rsync_queue'], states => ['active', 'inactive']});
+
+    for my $job (@{$results->{jobs}}) {
+        $jobs{$job->{id}} = _extend_job_info($job);
+    }
+
+    $self->render('ObsRsync_gru_index', jobs => \%jobs);
+}
+
+# this prepares fields for rendering
+sub _extend_job_info {
+    my ($job) = @_;
+
+    my $created_at = $job->{created};
+    $created_at = strftime("%Y-%m-%d %H:%M:%S %z", localtime($created_at)) if $created_at;
+
+    my $started_at = $job->{started};
+    $started_at = strftime("%Y-%m-%d %H:%M:%S %z", localtime($started_at)) if $started_at;
+
+    my $args = $job->{args};
+    if (ref $args eq 'HASH' && scalar(%$args) == 1 && $args->{project}) {
+        $args = $args->{project};
+    }
+    else {
+        $args = dump($args);
+    }
+    my $info = {
+        id       => $job->{id},
+        task     => $job->{task},
+        args     => $args,
+        created  => $created_at,
+        started  => $started_at,
+        priority => $job->{priority},
+        retries  => $job->{retries},
+        state    => $job->{state},
+        notes    => dump($job->{notes}),
+    };
+    return $info;
+}
+
+sub run {
```

Comment:
This method seems unused now.

Reply:
It's route created in both plugin_r https://github.com/os-autoinst/openQA/pull/2225/files#diff-69bf2c9a71fe44e7122df05bb819f8deR50 and plugin_api_r https://github.com/os-autoinst/openQA/pull/2225/files#diff-69bf2c9a71fe44e7122df05bb819f8deR66

Reply:
But this is not a controller.

Reply:
It is, at least the plan was that it will be a Controller and it works as Controller.
`Folders` and `Gru` have common checks, so the plan was that they are derived from `ObsRsync::Controller`, which has base class  `Mojolicious::Controller`.
https://github.com/os-autoinst/openQA/pull/2225/files#diff-7f23625de2ba6298272df078fcf961efR18

Reply:
Then make a real controller namespace containing two controller classes.

Reply:
I don't find this requirement "class in wrong place" as severe problem which can block this ticket. Please accept this PR and shift around classes the way you think is correct - otherwise it will never be finished. Of course I am glad to address real problems if you spot any.

Reply:
Unfortunately too much technical debt has been accumulated in openQA over the past few years. Mostly due to lack of proper reviews and enforcement of a consistent coding style. We should try our best not to add to that.

Reply:
It is not about accumulating new debt - it is about efficiency of reviews. It would be much more time consuming for you to shift classes around than trying to explain me how to deal with every challenge I meet trying to satisfy requirements that I don't full understand and cannot find similar cases in existing code.

Reply:
The lack of existing cases is a valid concern, and i'd be willing to clean up the classes in a followup PR if the other concerns are addresses.